### PR TITLE
Reduce allocations during check for duplicate analyzers

### PR DESCRIPTION
**Pull request by ToddGrun**
The HasDuplicates extension method always creates a HashSet of unspecified size when the number of entries being checked exceeds 2. As the CompilationWithAnalyzers Verify methods call this on large collections, there would be a significant amount of resizing during the checking for duplicates. I didn't want to change the extension method to use pooling on an unknown type parameter, and didn't want to change it to specify a capacity as I don't know the characteristics of other callers into the method.

Instead, I duplicated the HasDuplicates method directly into CompilationWithAnalyzers and modified it to use a pooled hashset (there are other uses of PooledHashSet<DiagnosticAnalyzer> in the code, so this doesn't introduce a new pool)

Old memory allocations:
image

New memory allocations:
image

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:145]: remove HasDuplicates if no longer used.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:165]:
```suggestion
                    var foundDuplicate = false;
```

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs
Comment (ToddGrun) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:145]: Still used in two other places

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs
Comment (333fred) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:152]: Why can't we just update `HasDuplicates()` to use a pooled hashset?

Comment (ToddGrun) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:152]: The concern I had with doing that was that it could create a pool that the caller didn't intentionally mean to create.

Comment (333fred) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:152]: Most of our extensions like this do, I don't think this would be any different.

Comment (ToddGrun) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:152]: This is slightly different in that the default pooledhashset pool assumes a default IEqualityComparer. One of the remaining callers of HasDuplicates passes in a custom comparer.

I could create a 2nd version of HasDuplicates in the extensions that doesn't take in the comparer and can use the default pooledhashset if you prefer.

Comment (ToddGrun) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: Went ahead and made the change such that HasDuplicates has two extension methods, depending on whether a custom comparer is used. If one isn't used, then the code utilizes PooledHashSet.

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:896]: nit, have this call into the other if comparer is null or EqualityComparer<T>.Default.

Comment (ToddGrun) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:896]: I'm going to hold off on making that change unless you or someone else feels strongly for it. My thought it that it adds a bit of complication to this code to better improve perf for a caller that simply shouldn't pass in the 2nd param.

Comment (333fred) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:896]: Agreed with Todd here. It doesn't seem worth the code complexity.

**Review by 333fred — APPROVED** [review]

**Comment by ToddGrun** [comment]
@dotnet/roslyn-compiler for a second review

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:896]: >HasDuplicates [](http://example.com/codeflow?start=29&length=13)

The difference in usage of the poll between the two overloads bothers me somewhat. Consider documenting the fact in a doc comment, recommending to use the other overload when the comparer used is the default one.

**Pull request merged**
